pf/internal/defaults/defaults.go

```diff
@@ -0,0 +1,293 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package defaults
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/unstable/propertyvalue"
+)
+
+type ApplyDefaultInfoValuesArgs struct {
+	// Required. The configuration property map to extend with defaults.
+	PropertyMap resource.PropertyMap
+
+	// Toplevel schema map for the resource, data source or provider.
+	TopSchemaMap shim.SchemaMap
+
+	// Toplevel SchemaInfo configuration matching TopSchemaMap.
+	TopFieldInfos map[string]*tfbridge.SchemaInfo
+
+	// Optional. Note that ResourceInstance need not be specified for Invoke or Configure processing.
+	ResourceInstance *tfbridge.PulumiResource
+
+	// Optional. If known, these are the provider-level configuration values, to support DefaultInfo.Config.
+	ProviderConfig resource.PropertyMap
```

Comment:
`PropertyMap` is required. `ResourceInstance` and `ProviderConfig` are optional. What about `TopSchemaMap` and `TopFieldInfos`?

Why are they `Top*`?

Reply:
Inside the module I was trying not to get confused with the recursive shim.SchemaMap and map[string]*tfbridge.SchemaInfo being looked up for the current object context (at path /x/y/z), vs the top-level (at path /) schema from which the lookups start. It's not really relevant for the consumers of the module though so I can edit that out.